lib/fika/compiler/type_checker/function_dependencies.ex

```diff
@@ -0,0 +1,31 @@
+defmodule Fika.Compiler.TypeChecker.FunctionDependencies do
+  @moduledoc """
+  This module keeps track of dependencies between functions.
+
+  The main usage for this is so we can avoid deadlocks in the
+  parallel compiler when there's recursion involved.
+  """
+
+  use Agent
+
+  def start_link(_) do
+    Agent.start_link(fn -> MapSet.new() end, name: __MODULE__)
+  end
+
+  @spec set(f :: String.t() | nil, g :: String.t() | nil) :: :ok | {:error, :cycle_encountered}
+  def set(f, g)
+
+  def set(f, g) when is_nil(f) or is_nil(g) do
+    :ok
+  end
+
+  def set(f, g) do
```

Comment:
Instead of mapping `{f, g}`, ideally we should keep track of all the indirect dependencies so we can handle cases like `a -> b -> c -> a`

A data structure like the one below would solve this.

`%{a: MapSet([b, c]), b: MapSet([c, a]), c: MapSet([b, a]}`